### PR TITLE
Add toString and getMessage method

### PR DESCRIPTION
**Pull request by hchoura**

### CODE REVIEW

File: src/foam/core/CompoundContextAgency.js [hunk 1]
```diff
@@ -42,7 +43,7 @@ for ( Runnable agent : getAgents() ) {
     e.add(t);
   }
 }
-e.maybeThrow();`
+e.maybeThrow((AppConfig) getX().get("appConfig"));`
```
Comment (jlhughes) [inline src/foam/core/CompoundContextAgency.js]: replace this with the context please. 
```suggestion
e.maybeThrow(getX());`
```

File: src/foam/nanos/ruler/RuleEngine.java [hunk 2]
```diff
@@ -84,6 +84,7 @@ public void execute(List<Rule> rules, FObject obj, FObject oldObj) {
       compoundAgency.execute(x_);
     } catch (Exception e) {
       logger.error(e.getMessage(), e);
+      throw e;
```
Comment (jlhughes) [inline src/foam/nanos/ruler/RuleEngine.java]: Who is catching this?

File: src/foam/core/CompoundException.java [hunk 3]
```diff
@@ -7,19 +7,31 @@
...
+  public String toString() {
+    return "CompoundException [exceptions_=" + exceptions_ + "]";
+  }
+
+  public String getMessage() {
+    return super.getMessage();
+  }
+
+  public void maybeThrow(AppConfig appConfig) {
+    if ( Mode.TEST == appConfig.getMode() ) {
```
Comment (jlhughes) [inline src/foam/core/CompoundException.java]:
```suggestion
    if ( Mode.TEST == ((AppConfig) x.get("appConfig")).getMode() ) {
```

File: src/foam/core/CompoundException.java [hunk 4]
```diff
@@ -7,19 +7,31 @@
...
-      throw this;
+  public String toString() {
+    return "CompoundException [exceptions_=" + exceptions_ + "]";
+  }
+
+  public String getMessage() {
+    return super.getMessage();
+  }
+
+  public void maybeThrow(AppConfig appConfig) {
```
Comment (jlhughes) [inline src/foam/core/CompoundException.java]:
```suggestion
  public void maybeThrow(X x) {
```

File: src/foam/core/CompoundException.java [hunk 5]
```diff
@@ -7,19 +7,31 @@
...
+  }
+
+  public String getMessage() {
+    return super.getMessage();
+  }
+
+  public void maybeThrow(X x) {
+    if ( Mode.TEST == ((AppConfig) x.get("appConfig")).getMode() ) {
+      if ( exceptions_.size() != 0 ) {
+        throw this;
```
Comment (jlhughes) [inline src/foam/core/CompoundException.java]: maybeThrow should always through when exceptions_.size() > 0. 
The test for TEST was for you generating a System.err  or printStackTrace that you had before.

File: src/foam/core/CompoundException.java [hunk 6]
```diff
@@ -7,19 +7,31 @@
...
 
-  public void maybeThrow() {
-    if ( exceptions_.size() != 0 ) {
-      throw this;
+  public String toString() {
+    return "CompoundException [exceptions_=" + exceptions_ + "]";
+  }
+
+  public String getMessage() {
+    return super.getMessage();
```
Comment (jlhughes) [inline src/foam/core/CompoundException.java]: How is this useful?

File: src/foam/core/CompoundException.java (hunk 5 above)
Comment (hchoura) [reply src/foam/core/CompoundException.java]: @kgrgreer ask to keep it

File: src/foam/core/CompoundContextAgency.js [hunk 7]
```diff
@@ -13,6 +13,7 @@ foam.CLASS({
   ],
 
   javaImports: [
+    'foam.nanos.app.AppConfig',
```
Comment (kgrgreer) [inline src/foam/core/CompoundContextAgency.js]: Why do we need this line?

File: src/foam/core/CompoundException.java (hunk 6 above)
Comment (kgrgreer) [reply src/foam/core/CompoundException.java]: It should create a StringBuilder and concatenate the getMessage() of each of the exception_s.

File: src/foam/core/CompoundException.java [hunk 8]
```diff
@@ -7,19 +7,31 @@
...
+  }
+
+  public void maybeThrow(X x) {
+    if ( Mode.TEST == ((AppConfig) x.get("appConfig")).getMode() ) {
```
Comment (kgrgreer) [inline src/foam/core/CompoundException.java]: This isn't required. Should always throw if size() !- 0

File: src/foam/core/CompoundException.java (hunk 5 above)
Comment (jlhughes) [reply src/foam/core/CompoundException.java]: @hchoura yes, keep the test and generate the standout call you need. but now this only rethrows in Test mode, that is incorrect.

**Comment by jlhughes** [comment]
@hchoura @lchanmann has done some of this in https://github.com/foam-framework/foam2/pull/3494

**Pull request closed without merging**